### PR TITLE
Calendar

### PR DESCRIPTION
**Pull request by jimmarino**
This is a work-in-progress to support import/export of Office calendar data.

One thing to note is that I had to update the Importer SPI signature to take the jobId:

ImportResult importItem(String jobId, A authData, T data);

This is needed so importers can persist scratch data (e.g. Calendar exported id to imported id mappings) associated with a job in the JobStore. Currently it's a String but I think we need to change that to UUID in a subsequent pull request.

### CODE REVIEW

File: extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/MicrosoftAuthDataGenerator.java [hunk 1]
```diff
@@ -51,7 +51,7 @@ public MicrosoftAuthDataGenerator(String redirectPath, Supplier<String> clientId
     public AuthFlowConfiguration generateConfiguration(String callbackBaseUrl, String id) {
         // constructs a request for the Microsoft Graph authorization code.
         String redirectUrl = callbackBaseUrl + redirectPath;
-        String queryPart = constructAuthQueryPart(redirectUrl, id, "user.read", "mail.read", "Contacts.ReadWrite");
+        String queryPart = constructAuthQueryPart(redirectUrl, id, "user.read", "mail.read", "Contacts.ReadWrite", "Calendars.ReadWrite");
```
Comment (holachuy) [inline extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/MicrosoftAuthDataGenerator.java]: Merging in the new AuthMode can be used in determining which permissions to allow 
https://github.com/google/data-portability/commit/d8bb624c43f1ba2a24902be984f5523616f9e38e#diff-6e743d98353302c9aac7752f60ce39e2

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+    private final OkHttpClient client;
+    private final ObjectMapper objectMapper;
+    private final TransformerService transformerService;
+
+    public MicrosoftCalendarExporter(OkHttpClient client, ObjectMapper objectMapper, TransformerService transformerService) {
+        this("https://graph.microsoft.com/", client, objectMapper, transformerService);
+    }
+
+    public MicrosoftCalendarExporter(String baseUrl, OkHttpClient client, ObjectMapper objectMapper, TransformerService transformerService) {
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java]: Is this only for testing? If so, we can add @ VisibleForTesting

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    private String calculateEventsUrl(String eventId) {
+        return baseUrl + String.format(EVENTS_URL, eventId);
+    }
+
+    @Override
+    public ExportResult<CalendarContainerResource> export(TokenAuthData authData, ExportInformation exportInformation) {
+        throw new UnsupportedOperationException();
+    }
+
+    private Request.Builder getBuilder(String url, TokenAuthData authData) {
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java]: private static

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/common/RequestHelper.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        } catch (IOException e) {
+            // TODO log
+            e.printStackTrace();
+            return new BatchResponse(new ImportResult(ImportResult.ResultType.ERROR, "Error batching request: " + e.getMessage()));
+        }
+    }
+
+    public static class BatchResponse {
+        private final ImportResult result;
+        private final List<Map<String, Object>> batchRespoonse;
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/common/RequestHelper.java]: typo  batchRespoonse

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/contacts/MicrosoftContactsImporter.java [hunk 5]
```diff
@@ -32,32 +28,34 @@
...
+import static org.dataportabilityproject.transfer.microsoft.common.RequestHelper.createRequest;
 
 /**
  * Performs a batch import of contacts using the Microsoft Graph API. For details see: https://developer.microsoft.com/en-us/graph/docs/concepts/json_batching.
  */
 public class MicrosoftContactsImporter implements Importer<TokenAuthData, ContactsModelWrapper> {
-    private static final String BATCH_URL = "https://graph.microsoft.com/beta/$batch";
     private static final String CONTACTS_URL = "me/contacts"; //must be relative for batch operations
 
+    private String baseUrl;
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/contacts/MicrosoftContactsImporter.java]: final

File: portability-worker/src/main/java/org/dataportabilityproject/worker/PortabilityInMemoryTransferCopier.java [hunk 6]
```diff
@@ -101,7 +102,7 @@ private void copy(
     logger.debug("Finished export, results: {}", exportResult);
 
     logger.debug("Starting import");
-    importer.importItem(importAuthData, exportResult.getExportedData());
+    importer.importItem("1", importAuthData, exportResult.getExportedData());
```
Comment (holachuy) [inline portability-worker/src/main/java/org/dataportabilityproject/worker/PortabilityInMemoryTransferCopier.java]: TODO, use job Id?

File: extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarModelTransformerTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Before
+    public void setUp() {
+        transformer = new ToCalendarModelTransformer();
+        mapper = new ObjectMapper();
+        context = new TestTransformerContext();
+    }
+
+    private static final String JSON = "{\n" +
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarModelTransformerTest.java]: Can we change the name to something more descriptive, e.g. SAMPLE_CALENDAR_EVENT or add a comment to describe

File: extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarEventModelTransformerTest.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    @Before
+    public void setUp() {
+        transformer = new ToCalendarEventModelTransformer();
+        mapper = new ObjectMapper();
+        context = new TestTransformerContext();
+    }
+
+    private static final String JSON =
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarEventModelTransformerTest.java]: We can probably use a more descriptive variable name or add comment

File: extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/helper/MockJobStore.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * An implementation for testing.
+ */
+public class MockJobStore implements JobStore {
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/helper/MockJobStore.java]: Can we eventually use the in memory job store when it's ported?

cc : @rtannenbaum

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarEventModelTransformer.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+ */
+public class ToCalendarEventModelTransformer implements BiFunction<Map<String, Object>, TransformerContext, CalendarEventModel> {
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public CalendarEventModel apply(Map<String, Object> event, TransformerContext context) {
+        if (!"singleInstance".equals(event.get("type"))) {   // support single instances for now;recurring events later
+            return null;
+        }
+        String calendarId = context.getProperty(CALENDAR_ID);
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarEventModelTransformer.java]: Do we always know we have a calendar_id in the context by this point?

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java (hunk 2 above)
Comment (jimmarino) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java]: OK

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java (hunk 3 above)
Comment (jimmarino) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java]: Do you think that should be static? It's a private method only called by other non-static members of that class

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/common/RequestHelper.java (hunk 4 above)
Comment (jimmarino) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/common/RequestHelper.java]: OK

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/contacts/MicrosoftContactsImporter.java (hunk 5 above)
Comment (jimmarino) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/contacts/MicrosoftContactsImporter.java]: ok

File: portability-worker/src/main/java/org/dataportabilityproject/worker/PortabilityInMemoryTransferCopier.java (hunk 6 above)
Comment (jimmarino) [reply portability-worker/src/main/java/org/dataportabilityproject/worker/PortabilityInMemoryTransferCopier.java]: OK

File: extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarModelTransformerTest.java (hunk 7 above)
Comment (jimmarino) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarModelTransformerTest.java]: Sure

File: extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarEventModelTransformerTest.java (hunk 8 above)
Comment (jimmarino) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarEventModelTransformerTest.java]: OK

File: extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/helper/MockJobStore.java (hunk 9 above)
Comment (jimmarino) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/helper/MockJobStore.java]: Probably can

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarEventModelTransformer.java (hunk 10 above)
Comment (jimmarino) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarEventModelTransformer.java]: Yes. I think this will come up for other transformations as well that require some form of context data to complete a transformation. I don't think we want to bake that into a context signature since this type of data can vary widely.

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java (hunk 3 above)
Comment (holachuy) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java]: It may be helpful to readers of the method to indicate that it doesn't depend on the state of member variables in the class versus trying to discern from reading the contents.

File: extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/helper/MockJobStore.java (hunk 9 above)
Comment (holachuy) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/helper/MockJobStore.java]: Maybe we can add a comment or an issue for a mockjobstore testing articfact

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarEventModelTransformer.java (hunk 10 above)
Comment (holachuy) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarEventModelTransformer.java]: Agreed, I'm just wondering what if we get an event without the outer calendar information, do we know the ID?

File: extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/helper/MockJobStore.java (hunk 9 above)
Comment (rtannenbaum) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/helper/MockJobStore.java]: It's ported! If you want to give it a shot: https://github.com/google/data-portability/blob/dc85c0e060cd3f9ba322e482ad7085ea6d62ba41/extensions/cloud/portability-cloud-local/src/main/java/org/dataportabilityproject/cloud/local/InMemoryKeyValueStore.java

File: extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/integration/MicrosoftCalendarExportTest.java [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+        token = new TokenAuthData("token123");
+        server = new MockWebServer();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        server.shutdown();
+    }
+
+    private static final String CALENDARS_RESPONSE = "{" +
```
Comment (rtannenbaum) [inline extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/integration/MicrosoftCalendarExportTest.java]: can these be read from a json file so we don't need all the string concatenation? fine to leave a TODO for separate PR. here & below

File: portability-spi-transfer/src/main/java/org/dataportabilityproject/spi/transfer/provider/Importer.java [hunk 12]
```diff
@@ -26,10 +26,11 @@
     /**
      * Imports data.
      *
+     * @param jobId the current job id
      * @param authData authentication information
      * @param data the data
      * @return the operation result
      */
-    ImportResult importItem(A authData, T data); // REVIEW: The original throws IOException. Continue to use or return as part of the result?
+    ImportResult importItem(String jobId, A authData, T data); // REVIEW: The original throws IOException. Continue to use or return as part of the result?
```
Comment (rtannenbaum) [inline portability-spi-transfer/src/main/java/org/dataportabilityproject/spi/transfer/provider/Importer.java]: Can this be a UUID? I made it so we use UUID instead of String most places, see https://github.com/google/data-portability/pull/133

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarEventModelTransformer.java (hunk 10 above)
Comment (jimmarino) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarEventModelTransformer.java]: Ah OK. In the MS API the exporter always exports Calendar containers first.

File: extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/integration/MicrosoftCalendarExportTest.java (hunk 11 above)
Comment (jimmarino) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/integration/MicrosoftCalendarExportTest.java]: I'd prefer the unit tests don't access the file system if possible but can do either way.

File: portability-spi-transfer/src/main/java/org/dataportabilityproject/spi/transfer/provider/Importer.java (hunk 12 above)
Comment (jimmarino) [reply portability-spi-transfer/src/main/java/org/dataportabilityproject/spi/transfer/provider/Importer.java]: I think so. That should be changed in a separate pull request,

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java (hunk 3 above)
Comment (jimmarino) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java]: Sure we can do this

File: extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/helper/MockJobStore.java (hunk 9 above)
Comment (jimmarino) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/helper/MockJobStore.java]: OK I'll try that in a subsequent pull request.

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+        List<CalendarModel> calendarModels = new ArrayList<>();
+        try (Response graphResponse = client.newCall(calendarsBuilder.build()).execute()) {
+            ResponseBody body = graphResponse.body();
+            if (body == null) {
+                return new ExportResult<>(ExportResult.ResultType.ERROR, "Error retrieving contacts: response body was null");
+            }
+            String graphBody = new String(body.bytes());
+            Map graphMap = objectMapper.reader().forType(Map.class).readValue(graphBody);
+
+            // String nextLink = (String) graphMap.get(ODATA_NEXT);
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java]: Does this need to be uncommented? Otherwise, perhaps add a TODO?

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+
+            try (Response graphResponse = client.newCall(eventsBuilder.build()).execute()) {
+                ResponseBody body = graphResponse.body();
+                if (body == null) {
+                    return new ExportResult<>(ExportResult.ResultType.ERROR, "Error retrieving calendar: response body was null");
+                }
+                String graphBody = new String(body.bytes());
+                Map graphMap = objectMapper.reader().forType(Map.class).readValue(graphBody);
+
+                // String nextLink = (String) graphMap.get(ODATA_NEXT);
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java]: Likewise TODO here

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java (hunk 3 above)
Comment (holachuy) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/calendar/MicrosoftCalendarExporter.java]: Not sure if you wanted to update it now or not.
